File: backend/GameApp.Application/Services/EnemyServices/EnemyCreateService.cs
using GameApp.Application.UseCases.EnemyUseCases;
using GameApp.Domain.Entities;
using GameApp.Domain.Repositories;

namespace GameApp.Application.Services.EnemyServices;


public class EnemyCreateService : EnemyCreateUseCase
{
    private readonly IEnemyRepository _repo;

    public EnemyCreateService(IEnemyRepository repo) => _repo = repo;

    public async Task<Enemy?> CreateEnemy(Enemy Enemy)
    {
        // Enemy name must be unique in the collection
        var existingEnemy = await _repo.FetchByName(Enemy.GetName());
        if (existingEnemy is not null)
        {
            Console.WriteLine($"Enemy name {Enemy.GetName()} exist");
            return null;
        }
        return await _repo.SaveAsync(Enemy);
    }
}

File: frontend/gameapp-frontend/src/application/services/enemy-services/enemy-create-service.ts
import { Enemy } from "../../../domain/entities/enemy";
import type { IEnemyCreateUseCase } from "../../usecases/enemy-use-cases/enemy-create-use-case";
import type { IEnemyRepository } from "../../repositories/ienemy-repository";

export class EnemyCreateService implements IEnemyCreateUseCase {
    constructor(private enemyRepository: IEnemyRepository) {}

    async createEnemy(enemy: Enemy): Promise<Enemy> {
        return await this.enemyRepository.save(enemy);
    }
}
